python/lsst/ctrl/mpexec/mpGraphExecutor.py

```diff
@@ -157,74 +293,85 @@ def _executeQuantaMP(self, iterable, butler):
-            if res.ready():
-                _LOG.debug("Result %d is ready", qid)
+        while jobs.pending() or jobs.running():
+
+            _LOG.debug("#pendingJobs: %s", len(jobs.pending()))
+            _LOG.debug("#runningJobs: %s", len(jobs.running()))
+
+            # See if any jobs have finished
+            for job in jobs.running():
+                proc = job.process
+                if not proc.is_alive():
+                    _LOG.debug("finished: %s", job)
+                    # finished
+                    if proc.exitcode == 0:
+                        job.state = JobState.FINISHED
+                        _LOG.debug("success: %s", job)
+                    else:
+                        job.state = JobState.FAILED
+                        _LOG.debug("failed: %s", job)
+                        if self.failFast:
+                            raise MPGraphExecutorError(
+                                f"Task {job} failed, exit code={proc.exitcode}."
+                            )
+                        else:
+                            _LOG.error(
+                                "Task %s failed; processing will continue for remaining tasks.", job
+                            )
+                else:
+                    # check for timeout
+                    now = time.time()
+                    if now - job.started > self.timeout:
+                        job.state = JobState.TIMED_OUT
+                        _LOG.debug("Terminating job %s due to timeout", job)
+                        job.stop()
+                        if self.failFast:
+                            raise MPTimeoutError(f"Timeout ({self.timeout} sec) for task {job}.")
+                        else:
+                            _LOG.error(
+                                "Timeout (%s sec) for task %s; task is killed, processing continues "
+                                "for remaining tasks.", self.timeout, job
+                            )
+
+            # see if we can start more jobs
+            for job in jobs.pending():
+
+                # check all dependencies
+                if job.qdata.dependencies & jobs.failedIds():
+                    # upstream job has failed, skipping this
+                    job.state = JobState.FAILED_DEP
+                    _LOG.error("Upstream job failed for task %s, skipping this task.", job)
+                elif job.qdata.dependencies <= jobs.finishedIds():
+                    # all dependencies have completed, can start new job
+                    if len(jobs.running()) < self.numProc:
+                        _LOG.debug("Sumbitting %s", job)
+                        job.start(butler, self.quantumExecutor)
+
+            # Here we want to wait until one of the running jobs completes
+            # but multiprocessing does not provide an API for that, for now
+            # just sleep a little bit and go back to the loop.
+            if jobs.running():
+                time.sleep(0.1)
```

Comment:
I can't help pondering asyncio in cases like this where we are sleeping and looping.

Reply:
I'm not very proud of this two lines of code, it's stupid and inefficient. We could certainly do fancier things here, in the context of pipetask it probably does not matter very much, but generally I agree we want to make things behaving better. Maybe something for later ticket together with finally getting rid of `multiprocessing` stuff.